frame/vesting/src/lib.rs

```diff
@@ -61,41 +60,139 @@ use frame_support::traits::{
+	}
+
+	impl<Balance: AtLeast32BitUnsigned + Copy, BlockNumber: AtLeast32BitUnsigned + Copy>
+		VestingInfo<Balance, BlockNumber>
+	{
+		/// Instantiate a new `VestingInfo` and validate parameters
+		pub fn try_new<T: Config>(
+			locked: Balance,
+			per_block: Balance,
+			starting_block: BlockNumber,
+		) -> Result<VestingInfo<Balance, BlockNumber>, Error<T>> {
+			Self::validate_params(locked, per_block, starting_block)?;
+			let per_block = if per_block > locked { locked } else { per_block };
+			Ok(VestingInfo { locked, per_block, starting_block })
+		}
+
+		/// Validate parameters for `VestingInfo`.
+		pub fn validate_params<T: Config>(
+			locked: Balance,
+			per_block: Balance,
+			_starting_block: BlockNumber,
+		) -> Result<(), Error<T>> {
+			ensure!(!locked.is_zero() && !per_block.is_zero(), Error::<T>::InvalidScheduleParams);
+			let min_transfer: u32 = T::MinVestedTransfer::get().try_into().unwrap_or(u32::MAX);
+			let min_transfer = Balance::from(min_transfer);
+			ensure!(locked >= min_transfer, Error::<T>::AmountLow);
+			Ok(())
+		}
+
+		/// Instantiate a new `VestingInfo` without param validation. Useful for
+		/// mocking bad inputs in testing.
+		pub fn unsafe_new(
+			locked: Balance,
+			per_block: Balance,
+			starting_block: BlockNumber,
+		) -> VestingInfo<Balance, BlockNumber> {
+			VestingInfo { locked, per_block, starting_block }
+		}
+
+		/// Locked amount at genesis.
+		pub fn locked(&self) -> Balance {
+			self.locked
+		}
+
+		/// Amount that gets unlocked every block after `starting_block`.
+		pub fn per_block(&self) -> Balance {
+			self.per_block
+		}
+
+		/// Starting block for unlocking(vesting).
+		pub fn starting_block(&self) -> BlockNumber {
+			self.starting_block
+		}
+
+		/// Amount locked at block `n`.
+		pub fn locked_at<BlockNumberToBalance: Convert<BlockNumber, Balance>>(
+			&self,
+			n: BlockNumber,
+		) -> Balance {
+			// Number of blocks that count toward vesting
+			// Saturating to 0 when n < starting_block
+			let vested_block_count = n.saturating_sub(self.starting_block);
+			let vested_block_count = BlockNumberToBalance::convert(vested_block_count);
+			// Return amount that is still locked in vesting
+			let maybe_balance = vested_block_count.checked_mul(&self.per_block);
+			if let Some(balance) = maybe_balance {
+				self.locked.saturating_sub(balance)
+			} else {
+				Zero::zero()
+			}
+		}
+
+		/// Block number at which the schedule ends
+		pub fn ending_block<BlockNumberToBalance: Convert<BlockNumber, Balance>>(&self) -> Balance {
+			let starting_block = BlockNumberToBalance::convert(self.starting_block);
+			let duration = if self.per_block > self.locked {
+				// If `per_block` is bigger than `locked`, the schedule will end
+				// the block after starting
+				1u32.into()
+			} else if self.per_block.is_zero() {
+				// Check for div by 0 errors, which should only be from legacy
+				// vesting schedules since new ones are validated for this.
+				self.locked
```

Comment:
Using `self.locked` is accurate if and only if 1 block is equivalent to 1 planck. The existence of `BlockNumberToBalance` suggests that we cannot depend on that equivalence.

Reply:
Yup good point, I suppose we need to saturate at the max block number. 

However if this edge case is real we will have some fairly useless schedules in storage for the life of the chain. So we may want to have more discussion on what to do if we get here and the schedule still has a `per_block` of 0. With the new validation checks this should not happen, but its entirely possible that runtimes using this pallet have zero values. An alternative would just be deleting all those schedules in a migration, which would effectively burn funds

Reply:
I think this should return infinity aka. `Bounded::max_value()` or better, an error.

Reply:
I created an error for schedules that have infinite duration:

089f5da417f42c0bbddb30147b4be818eddb2911